<img style="float: left; margin: 30px 15px 15px 15px;" src="https://pngimage.net/wp-content/uploads/2018/06/logo-iteso-png-5.png" width="300" height="500" /> 
    
    
### <font color='navy'> Simulación de procesos financieros. 

**Nombres:** Diego Alberto López Lazareno y Jesús Sebastián González Ortiz.

**Fecha:** 21 de septiembre del 2021.

**Expediente** : 722100 y 721621.
**Profesor:** Oscar David Jaramillo Zuluaga.
    
**Link Github**: https://github.com/SebasGlez2000/Tarea4_JSGonzalez_DALopez

# Tarea 4: Clase 7

## Enunciado de tarea 

**Ejercicio 1.** Como ejemplo simple de una simulación de Monte Carlo, considere calcular la probabilidad de una suma particular del lanzamiento de tres dados (cada dado tiene valores del uno al seis). Además cada dado tiene las siguientes carácterísticas: el primer dado no está cargado (distribución uniforme todos son equiprobables); el segundo y tercer dado están cargados basados en una distribución binomial con parámetros (`n=5, p=0.5` y `n=5, p=0.2`). Calcule la probabilidad de que la suma resultante sea 7, 14 o 18. 

**Ejercicio 2.** 
Premisas para la simulación:
- Negocio de alimentos que vende bebidas y alimentos.
- Negocio dentro del ITESO.
- Negocio en cafetería central.
- Tipo de clientes (hombres y mujeres).
- Rentabilidad del 60%.

Objetivo
Realizar una simulación estimado el tiempo medio que se tardaran los clientes en ser atendidos entre el horario de 6:30 a 1 pm. Además saber el consumo. 
**Analizar supuestos y limitantes**

Supuestos en simulación
Clasificación de clientes: 
- Mujer  = 1 $\longrightarrow$ aleatorio < 0.5
- Hombre = 0 $\longrightarrow$ aleatorio $\geq$ 0.5.

Condiciones iniciales:
- Todas las distrubuciones de probabilidad se supondrán uniformes.
- Tiempo de simulación: 6:30 am - 1:30pm $\longrightarrow$ T = 7 horas = 25200 seg.
- Tiempo de llegada hasta ser atendido: Min=5seg, Max=30seg.
- Tiempo que tardan los clientes en ser atendidos:
    - Mujer: Min = 1 min= 60seg, Max = 5 min = 300 seg
    - Hombre: Min = 40 seg, Max = 2 min= 120 seg
- Consumo según el tipo de cliente:
    - Mujer: Min = 30 pesos, Max = 100 pesos
    - Hombre: Min = 20 pesos, Max = 80 pesos

Responder las siguientes preguntas basados en los datos del problema:
1. ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.
2. ¿Cuál fue el consumo promedio de los hombres y mujeres?
3. ¿Cuál fue el número de personas atendidas por día?
4. ¿Cuál fue el tiempo de atención promedio?
5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?

In [1]:
# Librerías necesarias
import scipy.stats as st
import numpy as np
import pandas as pd 

### Ejercicio 1: (enunciado) 

Como ejemplo simple de una simulación de Monte Carlo, considere calcular la probabilidad de una suma particular del lanzamiento de tres dados (cada dado tiene valores del uno al seis). Además cada dado tiene las siguientes carácterísticas: el primer dado no está cargado (distribución uniforme todos son equiprobables); el segundo y tercer dado están cargados basados en una distribución binomial con parámetros (`n=5, p=0.5` y `n=5, p=0.2`). Calcule la probabilidad de que la suma resultante sea 7, 14 o 18. 

#### Código de solución Sebastián

In [2]:
# Código de solución estudiante 1


#### Código de solución Diego

In [3]:
# Código de solución estudiante 2

def ThreeDiceThrowing(p1 = "Probability of a single success for dice #2", 
                      p2 = "Probability of a single success for dice #3", size = "Throwings"):
    
    """ Returns the sum of a three dice throwing:
        The first dice follows a uniform distribution,
        The second dice follows a binomial distribution with a probability of a single success of p1,
        The third dice follows a binominal distribution with a probability of a single success of p2
        Size is the number of throwings that will be the same for all dices
    """
    np.random.seed(616)
    dice1 = np.random.randint(1, 7, size)
    dice2 = st.binom(5, p1, loc = 1).rvs(size = size)
    dice3 = st.binom(5, p2, loc = 1).rvs(size = size)
    
    return dice1 + dice2 + dice3


In [4]:
result = ThreeDiceThrowing(0.5, 0.2, 1000000)
n = len(result)
probabilities = list(map(lambda number : ("Probability that the resulting dice roll sum is " + str(number) + " : ", 
                                          len(result[result == number]) / n), [7, 14, 18]))
probabilities

[('Probability that the resulting dice roll sum is 7 : ', 0.127023),
 ('Probability that the resulting dice roll sum is 14 : ', 0.013619),
 ('Probability that the resulting dice roll sum is 18 : ', 0.0)]

La resolución de este ejercicio es relativamente sencilla, se propone el uso de la función `map` para ir calculando la probabilidad de que al tirar los $3$ dados con sus respectivas especificaciones arrojen como suma los números $7$, $14$ y $18$. Los resultados son los siguientes:

+ La probabilidad de que la suma resultante al lanzar los tres dados sea $7$ es de $12.70\%$.
+ La probabilidad de que la suma resultante al lanzar los tres dados sea $14$ es de $1.36\%$.
+ La probabilidad de que la suma resultante al lanzar los tres dados sea $18$ es de $0.00\%$.

### 2 Ejercicio de aplicación- Cafetería Central

Premisas para la simulación:
- Negocio de alimentos que vende bebidas y alimentos.
- Negocio dentro del ITESO.
- Negocio en cafetería central.
- Tipo de clientes (hombres y mujeres).
- Rentabilidad del 60%.

### Objetivo
Realizar una simulación estimado el tiempo medio que se tardaran los clientes en ser atendidos entre el horario de 6:30 a 1 pm. Además saber el consumo. 
**Analizar supuestos y limitantes**

## Supuestos en simulación
Clasificación de clientes: 
- Mujer  = 1 $\longrightarrow$ aleatorio < 0.5
- Hombre = 0 $\longrightarrow$ aleatorio $\geq$ 0.5.

Condiciones iniciales:
- Todas las distrubuciones de probabilidad se supondrán uniformes.
- Tiempo de simulación: 6:30 am - 1:30pm $\longrightarrow$ T = 7 horas = 25200 seg.
- Tiempo de llegada hasta ser atendido: Min=5seg, Max=30seg.
- Tiempo que tardan los clientes en ser atendidos:
    - Mujer: Min = 1 min= 60seg, Max = 5 min = 300 seg
    - Hombre: Min = 40 seg, Max = 2 min= 120 seg
- Consumo según el tipo de cliente:
    - Mujer: Min = 30 pesos, Max = 100 pesos
    - Hombre: Min = 20 pesos, Max = 80 pesos

Responder las siguientes preguntas basados en los datos del problema:
1. ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.
2. ¿Cuál fue el consumo promedio de los hombres y mujeres?
3. ¿Cuál fue el número de personas atendidas por día?
4. ¿Cuál fue el tiempo de atención promedio?
5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?

In [5]:
################## Datos del problema
d = 5
T = 25200
T_at_min = 5; T_at_max = 30
T_mujer_min = 60; T_mujer_max = 300
T_hombre_min = 40; T_hombre_max = 120
C_mujer_min = 30; C_mujer_max = 100
C_hombre_min = 20; C_hombre_max = 80

#### Código de solución Sebastián

In [6]:
# Código de solución estudiante 1


#### Código de solución Diego

In [7]:
# Código de solución estudiante 2

def CofeeShop(seed):
    # Semilla
    np.random.seed(seed)
    # Número de clientes
    n = 500
    # Almacenamiento de los datos
    data = pd.DataFrame(columns = ["Sex", "Waiting", "Customer Support", "Consumption"], index = range(n))
    
    ## ----------------------------------------------- Simulación ------------------------------------------------ ##
    # Sexo del cliente
    data["Sex"] = list(map(lambda rand : 1 if rand < 0.5 else 0, list(np.random.uniform(size = n))))
    # Tiempo de espera hasta ser atendido
    data["Waiting"] = np.random.uniform(T_at_min, T_at_max, size = n)
    # Tiempo que tarda en cliente en ser atendido
    data["Customer Support"] = list(map(lambda sex : np.random.uniform(T_mujer_min, T_mujer_max) if sex == 1 
                                        else np.random.uniform(T_hombre_min, T_hombre_max), list(data["Sex"])))
    # Consumo 
    data["Consumption"] = list(map(lambda sex : np.random.uniform(C_mujer_min, C_mujer_max) if sex == 1 
                                        else np.random.uniform(C_hombre_min, C_hombre_max), list(data["Sex"])))
    
    # Tiempo total de espera y atención por cliente
    data["Time Per Customer"] = data["Waiting"] + data["Customer Support"]
    # Tiempo acumulado en la simulación
    data["Accumulated Time"] = data["Time Per Customer"].cumsum()
    
    # Se filtra el DataFrame hasta donde alcanza el tiempo de simulación
    return data[data["Accumulated Time"] <= T]

# Simulación para 5 días 
seeds = [2001, 2010*616, 2061*8, 3001, 1984]
df = {"día " + str(i + 1) : CofeeShop(seeds[i]) for i in range(d)}

In [8]:
# Pregunta 1
ans1 = list(map(lambda sex : sum(list(map(lambda key : 
                                          sum(list(df[key][df[key]["Sex"] == sex]["Consumption"])), df.keys()))), [0, 1]))

print("Los gastos de los hombres en 5 días fueron de " + str(ans1[0]) + " pesos")
print("Los gastos de las mujeres en 5 días fueron de " + str(ans1[1]) + " pesos")

Los gastos de los hombres en 5 días fueron de 19791.653526913822 pesos
Los gastos de las mujeres en 5 días fueron de 27840.11059746975 pesos


In [9]:
# Pregunta 2
ans2 = list(map(lambda sex : np.mean(list(map(lambda key : 
                                          np.mean(list(df[key][df[key]["Sex"] == sex]["Consumption"])), df.keys()))), [0, 1]))

print("El consumo promedio de los hombres en 5 días fue de " + str(ans2[0]) + " pesos")
print("El consumo promedio de las mujeres en 5 días fue de " + str(ans2[1]) + " pesos")

El consumo promedio de los hombres en 5 días fue de 49.5124544776996 pesos
El consumo promedio de las mujeres en 5 días fue de 63.13278002244307 pesos


In [10]:
# Pregunta 3
ans3 = list(map(lambda key : "En el " + key + " se atendieron " + str(len(df[key])) + " personas", df.keys()))
ans3

['En el día 1 se atendieron 166 personas',
 'En el día 2 se atendieron 163 personas',
 'En el día 3 se atendieron 173 personas',
 'En el día 4 se atendieron 164 personas',
 'En el día 5 se atendieron 174 personas']

In [11]:
cols = ["Waiting", "Customer Support", "Time Per Customer"]
ans4 = list(map(lambda col : np.mean(list(map(lambda key : np.mean(df[key][col]), df.keys()))), cols))

print("El tiempo promedio de espera hasta ser atendido fue de " + str(ans4[0]) + " segundos")
print("El tiempo promedio de atención del cliente fue de " + str(ans4[1]) + " segundos")
print("El tiempo promedio total (de espera y de atención) fue de " + str(ans4[2]) + " segundos")

El tiempo promedio de espera hasta ser atendido fue de 17.317090458861458 segundos
El tiempo promedio de atención del cliente fue de 132.02246399061175 segundos
El tiempo promedio total (de espera y de atención) fue de 149.33955444947318 segundos


In [12]:
# Pregunta 5
ans5 = np.mean(list(map(lambda key : sum(df[key]["Consumption"]), df.keys())))

print("Las ventas promedio de la cafetería en 5 días de trabajo fueron de " + str(ans5) + " pesos")
print("La rentabilidad promedio de la cafetería en 5 días de trabajo fue de " + str(ans5 * 0.6) + " pesos")

Las ventas promedio de la cafetería en 5 días de trabajo fueron de 9526.352824876714 pesos
La rentabilidad promedio de la cafetería en 5 días de trabajo fue de 5715.811694926028 pesos


**1. ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.**

+ Los gastos totales de los hombres en los 5 días de trabajo fueron de $ \$ 19,791.65$.
+ Los gastos totales de las mujeres en los 5 días de trabajo fueron de $ \$ 27,840.11$.

**2. ¿Cuál fue el consumo promedio de los hombres y mujeres?**
+ El consumo promedio de los hombres fue de $ \$ 49.51$.
+ El consumo promedio de las mujeres fue de $ \$ 63.13$

**3. ¿Cuál fue el número de personas atendidas por día?**

+ En el día $1$ se atendieron $166$ personas.
+ En el día $2$ se atendieron $163$ personas.
+ En el día $3$ se atendieron $173$ personas.
+ En el día $4$ se atendieron $164$ personas.
+ En el día $5$ se atendieron $174$ personas.

**4. ¿Cuál fue el tiempo de atención promedio?**

+ El tiempo promedio de espera hasta ser atendido fue de $17.31$ segundos
+ El tiempo promedio de atención del cliente fue de $132.02$ segundos
+ El tiempo promedio total (de espera y de atención) fue de $149.33$ segundos

**5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?**

+ La ganancia promedio de la cafetería en 5 días de trabajo fueron de $ \$ 9,526.35$.
+ La rentabilidad promedio de la cafetería en 5 días de trabajo fue de $ \$ 5,715.81$.
